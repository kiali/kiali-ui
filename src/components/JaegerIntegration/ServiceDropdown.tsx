import * as React from 'react';
import { connect } from 'react-redux';
import { KialiAppState } from '../../store/Store';
import { FormSelect, FormSelectOption, FormSelectOptionGroup } from '@patternfly/react-core';
import * as Api from '../../services/Api';
import { PromisesRegistry } from '../../utils/CancelablePromises';
import { ServiceOverview } from '../../types/ServiceList';
import { style } from 'typestyle';

interface ServiceDropdownProps {
  disabled?: boolean;
  activeNamespaces: string[];
  service?: string;
  setService: (service: string) => void;
}

interface Service {
  value: string;
  label: string;
  disabled: boolean;
}

interface ServiceGroup {
  groupLabel: string;
  disabled: boolean;
  options: Service[];
}
interface ServiceDropdownState {
  servicesGroups: ServiceGroup[];
}

const serviceDropdown = style({ marginLeft: '-100px' });

export class ServiceDropdown extends React.PureComponent<ServiceDropdownProps, ServiceDropdownState> {
  constructor(props: ServiceDropdownProps) {
    super(props);
    this.state = { servicesGroups: [] };
    if (this.props.activeNamespaces.length > 0) {
      this.refreshServices(this.props.activeNamespaces);
    }
  }

  componentDidUpdate(prevProps: ServiceDropdownProps) {
    if (prevProps.activeNamespaces.sort().join(',') !== this.props.activeNamespaces.sort().join(',')) {
      this.refreshServices(this.props.activeNamespaces);
    }
  }

  refreshServices = (namespaces: string[]) => {
    if (namespaces.length === 0) {
      this.setState({ servicesGroups: [] });
    } else {
      const servicesPromises = namespaces.map(ns => Api.getServices(ns));
      const promises = new PromisesRegistry();
      promises
        .registerAll('services', servicesPromises)
        .then(responses => {
          const serviceList: ServiceGroup[] = [];
          responses.forEach(response => {
            const ns = response.data.namespace.name;
            const serviceGroup: ServiceGroup = { groupLabel: ns, disabled: false, options: [] };

            response.data.services.forEach((service: ServiceOverview) => {
              serviceGroup.options.push({ value: `${service.name}.${ns}`, label: service.name, disabled: false });
            });
            serviceList.push(serviceGroup);
          });
          this.setState({ servicesGroups: serviceList });
        })
        .catch(() => console.log('Error'));
    }
  };

  handleFocus = () => this.refreshServices(this.props.activeNamespaces);

  labelServiceDropdown = (items: number) => {
    if (this.props.activeNamespaces.length > 0) {
      if (items === 0) {
        return 'Select another namespace with services';
      }
      return 'Select a service';
    }
    return 'Select a namespace';
  };

  render() {
    const { disabled } = this.props;
    const { servicesGroups } = this.state;

    return (
      <FormSelect
        value={this.props.service}
        isDisabled={disabled || this.props.activeNamespaces.length === 0 || Object.keys(servicesGroups).length === 0}
        onFocus={() => this.handleFocus}
        onChange={this.props.setService}
        aria-label="FormSelect Input"
        className={serviceDropdown}
      >
        <FormSelectOption
          isDisabled={false}
          key={'help_test'}
          value={''}
          label={this.labelServiceDropdown(Object.keys(servicesGroups).length)}
        />
        {servicesGroups.map((group, index) => (
          <FormSelectOptionGroup isDisabled={group.disabled} key={index} label={group.groupLabel}>
            {group.options.map((option, i) => (
              <FormSelectOption isDisabled={option.disabled} key={i} value={option.value} label={option.label} />
            ))}
          </FormSelectOptionGroup>
        ))}
      </FormSelect>
    );
  }
}

const mapStateToProps = (state: KialiAppState) => {
  return {
    activeNamespaces: state.namespaces.activeNamespaces.map(ns => ns.name)
  };
};

const ServiceDropdownContainer = connect(mapStateToProps)(ServiceDropdown);

export default ServiceDropdownContainer;
