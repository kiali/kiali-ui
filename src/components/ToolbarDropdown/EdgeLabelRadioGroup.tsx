import * as React from 'react';
import PropTypes from 'prop-types';
import { style } from 'typestyle';
import * as _ from 'lodash';

import { GraphParamsType } from '../../types/Graph';
import { EdgeLabelMode } from '../../types/GraphFilter';

const EDGE_LABEL_MODES = EdgeLabelMode.getValues().reduce((map, edgeLabelMode) => {
  map[edgeLabelMode] = _.capitalize(_.startCase(edgeLabelMode));
  return map;
}, {});

type EdgeLabelRadioGroupProps = {
  graphParams: GraphParamsType;
  onEdgeChanged: PropTypes.func;
};

export class EdgeLabelRadioGroup extends React.Component<EdgeLabelRadioGroupProps> {
  constructor(props: EdgeLabelRadioGroupProps) {
    super(props);
  }

  onEdgeChanged = (event: any) => {
    const selectedEdge = event.target.value;
    this.setState({
      edgeValue: selectedEdge
    });
    this.props.onEdgeChanged(event);
  };

  render() {
    const radioButtonStyle = style({ marginLeft: 5 });

    const edgeItems = Object.keys(EDGE_LABEL_MODES).map((edgeLabelModeKey: any) => (
      <div key={edgeLabelModeKey}>
        <label className={radioButtonStyle}>
          <input
            type="radio"
            name="edge-label-radio-group"
            value={edgeLabelModeKey}
            checked={edgeLabelModeKey === this.props.graphParams.edgeLabelMode}
            onChange={this.onEdgeChanged}
          />
          <span className={radioButtonStyle}>{EDGE_LABEL_MODES[edgeLabelModeKey]}</span>
        </label>
      </div>
    ));

    return (
      <>
        <div>
          <label>Edge Labels:</label>
        </div>
        {edgeItems}
      </>
    );
  }
}

export default EdgeLabelRadioGroup;
