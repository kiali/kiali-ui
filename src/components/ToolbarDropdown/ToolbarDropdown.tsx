import * as React from 'react';
import { DropdownButton, MenuItem } from 'patternfly-react';
import PropTypes from 'prop-types';

type ToolbarDropdownProps = {
  disabled: boolean;
  nameDropdown?: string;
  initialValue: number | string;
  initialLabel: string | undefined;
  onClick: PropTypes.func;
  options: PropTypes.Map<string | number, string>;
};

type ToolbarDropdownState = {
  currentValue: number | string;
  currentName: string | undefined;
};

export class ToolbarDropdown extends React.Component<ToolbarDropdownProps, ToolbarDropdownState> {
  constructor(props: ToolbarDropdownProps) {
    super(props);
    this.state = {
      currentValue: props.initialValue,
      currentName: props.initialLabel
    };
  }

  onKeyChanged = (key: any) => {
    this.setState({ currentValue: key[0], currentName: key[1] });
    this.props.onClick(key[0]);
  };

  render() {
    return (
      <div className="form-group">
        {this.props.nameDropdown && <label style={{ paddingRight: '0.5em' }}>{this.props.nameDropdown}:</label>}
        <DropdownButton title={this.state.currentName} onSelect={this.onKeyChanged}>
          {this.props.options.map(r => (
            <MenuItem key={r[0]} active={r[0] === this.state.currentValue} eventKey={r}>
              {r[1]}
            </MenuItem>
          ))}
        </DropdownButton>
      </div>
    );
  }
}
