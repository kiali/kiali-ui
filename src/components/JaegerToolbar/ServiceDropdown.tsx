import * as React from 'react';
import { connect } from 'react-redux';
import { KialiAppState } from '../../store/Store';
import ToolbarDropdown from '../../components/ToolbarDropdown/ToolbarDropdown';
import { JaegerActions } from '../../actions/JaegerActions';
import { JaegerThunkActions } from '../../actions/JaegerThunkActions';
import { ThunkDispatch } from 'redux-thunk';
import { KialiAppAction } from '../../actions/KialiAppAction';

interface ServiceDropdownProps {
  disabled: boolean;
  activeNamespace: string;
  service: string;
  items: string[];
  refresh: (ns: string) => void;
  setService: (service: string) => void;
}

export class ServiceDropdown extends React.PureComponent<ServiceDropdownProps, {}> {
  constructor(props: ServiceDropdownProps) {
    super(props);
  }

  componentDidMount() {
    if (this.props.activeNamespace) {
      this.props.refresh(this.props.activeNamespace);
    }
  }

  componentDidUpdate(prevProps: ServiceDropdownProps) {
    if (this.props.activeNamespace !== prevProps.activeNamespace && this.props.activeNamespace) {
      this.props.refresh(this.props.activeNamespace);
    }
  }

  handleToggle = (isOpen: boolean) => isOpen && this.props.refresh(this.props.activeNamespace);

  labelServiceDropdown = (items: number) => {
    if (this.props.activeNamespace && this.props.activeNamespace !== 'all') {
      if (items === 0) {
        return 'Choose another namespace with services';
      }
      return 'Choose a service';
    }
    return 'Choose a namespace';
  };

  render() {
    const { disabled } = this.props;

    const items: { [key: string]: string } = this.props.items.reduce((list, item) => {
      list[item] = item;
      return list;
    }, {});

    return (
      <span style={{ marginLeft: '10px' }}>
        <ToolbarDropdown
          id="namespace-selector"
          disabled={disabled || Object.keys(items).length === 0}
          options={items}
          value={''}
          label={this.props.service || this.labelServiceDropdown(Object.keys(items).length)}
          useName={true}
          handleSelect={this.props.setService}
          onToggle={this.handleToggle}
        />
      </span>
    );
  }
}

const mapStateToProps = (state: KialiAppState) => {
  return {
    items: state.jaegerState.toolbar.services,
    disabled: state.jaegerState.toolbar.isFetchingService,
    activeNamespace: state.jaegerState.search.namespaceSelected,
    service: state.jaegerState.search.serviceSelected
  };
};

const mapDispatchToProps = (dispatch: ThunkDispatch<KialiAppState, void, KialiAppAction>) => {
  return {
    refresh: (ns: string) => {
      dispatch(JaegerThunkActions.asyncFetchServices(ns));
    },
    setService: (service: string) => {
      dispatch(JaegerActions.setService(service));
    }
  };
};

const ServiceDropdownContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(ServiceDropdown);

export default ServiceDropdownContainer;
