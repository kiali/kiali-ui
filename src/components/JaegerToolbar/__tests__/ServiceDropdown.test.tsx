import * as React from 'react';
import { shallow } from 'enzyme';
import { ServiceDropdown } from '../ServiceDropdown';
import ToolbarDropdown from '../../../components/ToolbarDropdown/ToolbarDropdown';

describe('NamespaceDropdown', () => {
  let wrapper, refresh, setService;
  beforeEach(() => {
    refresh = jest.fn();
    setService = jest.fn();
    wrapper = shallow(
      <ServiceDropdown
        refresh={refresh}
        setService={setService}
        service={''}
        activeNamespace={''}
        disabled={false}
        items={[]}
      />
    );
  });

  it('renders ServiceDropdown correctly without custom', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it('disable ToolbarDropwdown if no namespaces', () => {
    expect(wrapper.find(ToolbarDropdown).length).toEqual(1);
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['disabled']
    ).toBeTruthy();
    wrapper.setProps({ items: ['details', 'productpage'] });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['disabled']
    ).toBeFalsy();
    wrapper.setProps({ disabled: true });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['disabled']
    ).toBeTruthy();
  });

  it('value should be empty', () => {
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['value']
    ).toBe('');
  });

  it('label should be service', () => {
    const service = 'details';
    wrapper.setProps({ service: service });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['label']
    ).toBe(service);
  });

  it('options should be items', () => {
    const services = ['details', 'productpage'];
    const items: { [key: string]: string } = services.reduce((list, item) => {
      list[item] = item;
      return list;
    }, {});
    wrapper.setProps({ items: services });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['options']
    ).toEqual(items);
  });

  it('handleSelect should be setService', () => {
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['handleSelect']
    ).toBe(setService);
  });
});
