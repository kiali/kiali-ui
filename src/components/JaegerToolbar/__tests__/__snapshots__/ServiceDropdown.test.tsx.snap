// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceDropdown renders ServiceDropdown correctly without custom 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ServiceDropdown
    activeNamespace=""
    disabled={false}
    items={Array []}
    refresh={[MockFunction]}
    service=""
    setService={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ToolbarDropdown
        disabled={true}
        handleSelect={[MockFunction]}
        id="namespace-selector"
        label="Choose a namespace"
        onToggle={[Function]}
        options={Object {}}
        useName={true}
        value=""
      />,
      "style": Object {
        "marginLeft": "10px",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "disabled": true,
        "handleSelect": [MockFunction],
        "id": "namespace-selector",
        "label": "Choose a namespace",
        "onToggle": [Function],
        "options": Object {},
        "useName": true,
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ToolbarDropdown
          disabled={true}
          handleSelect={[MockFunction]}
          id="namespace-selector"
          label="Choose a namespace"
          onToggle={[Function]}
          options={Object {}}
          useName={true}
          value=""
        />,
        "style": Object {
          "marginLeft": "10px",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": true,
          "handleSelect": [MockFunction],
          "id": "namespace-selector",
          "label": "Choose a namespace",
          "onToggle": [Function],
          "options": Object {},
          "useName": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
